clippy_lints/src/mutable_debug_assertion.rs

```diff
@@ -0,0 +1,83 @@
+use crate::utils::{is_expn_of, match_qpath, span_note_and_lint};
+use rustc::hir::{Expr, ExprKind, Mutability, UnOp};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for function calls that mutate an object in `debug_assert!`, `debug_assert_eq!` and
+    /// `debug_assert_ne!` macros.
+    ///
+    /// **Why is this bad?** In release builds the code won't be executed and therefore can alter
+    /// the execution flow.
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// debug_assert_eq!(vec![3].pop(), Some(3));
+    /// // or
+    /// let mut s = String::from("foo");
+    /// debug_assert_eq!(s.remove(0), 'f');
+    /// ```
+    pub DEBUG_ASSERT_WITH_MUT_CALL,
+    correctness,
+    "`debug_assert!` and others will be ignored in release mode and therefore alter the execution flow. Replace it with non-mut function calls."
+}
+
+declare_lint_pass!(DebugAssertWithMutCall => [DEBUG_ASSERT_WITH_MUT_CALL]);
+
+fn check_for_mutable_call(e: &Expr) -> bool {
+    match &e.kind {
+        ExprKind::Call(func, args) => {
+            if let ExprKind::Path(path) = &func.kind {
+                if !match_qpath(path, &["new_v1"]) {
+                    for a in (*args).iter() {
+                        if let ExprKind::AddrOf(Mutability::MutMutable, _) = &a.kind {
+                            return true;
+                        }
+                    }
+                }
+            }
+        },
+        _ => (),
+    }
+
+    false
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for DebugAssertWithMutCall {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, e: &'tcx Expr) {
+        fn check_inner(e: &Expr) -> Option<&Expr> {
+            if let ExprKind::AddrOf(_, addrof) = &e.kind {
+                if check_for_mutable_call(&addrof) {
+                    return Some(addrof);
+                }
+            }
+            None
+        }
+
+        let mut emit_lint: Option<&Expr> = None;
+        if is_expn_of(e.span, "debug_assert").is_some() {
+            if let ExprKind::Unary(UnOp::UnNot, inner) = &e.kind {
+                if check_for_mutable_call(inner) {
+                    emit_lint = Some(&*inner);
+                }
+            }
+        } else if is_expn_of(e.span, "debug_assert_eq").is_some() {
+            emit_lint = check_inner(e);
+        } else if is_expn_of(e.span, "debug_assert_ne").is_some() {
+            emit_lint = check_inner(e);
+        }
+
+        if let Some(expr) = emit_lint {
+            span_note_and_lint(
+                cx,
+                DEBUG_ASSERT_WITH_MUT_CALL,
+                e.span,
+                "Do not call a mutable function inside of a debug_assert!",
+                expr.span,
+                "Invocation here",
```

Comment:
I think, this note can be removed, by directly using the `expr.span` above.

Reply:
So you want:
```
LL |     debug_assert_ne!(foo_mut(&mut 3), 1);
   |                      ^^^^^^^^^^^^^^^
```
only?

Reply:
Yes, I think that is enough.